src/core/datagram.c

```diff
@@ -0,0 +1,434 @@
+
+        if (Builder->Metadata->Flags.KeyType == QUIC_PACKET_KEY_0_RTT &&
+            !(SendRequest->Flags & QUIC_SEND_FLAG_ALLOW_0_RTT)) {
+            //
+            // Not allowed to send the datagram in 0-RTT.
+            //
+            return FALSE;
+        }
+
+        BOOLEAN HadRoomForDatagram =
+            QuicDatagramFrameEncodeEx(
+                SendRequest->Buffers,
+                SendRequest->BufferCount,
+                SendRequest->TotalLength,
+                &Builder->DatagramLength,
+                AvailableBufferLength,
+                (uint8_t*)Builder->Datagram->Buffer);
+        if (!HadRoomForDatagram && Builder->Metadata->FrameCount != 0) {
+            //
+            // Part of the packet was used already and we didn't have room to frame
+            // this datagram. We need to try again in an unused packet.
+            //
+            return FALSE;
+        }
+
+        if (Datagram->PrioritySendQueueTail == &SendRequest->Next) {
+            Datagram->PrioritySendQueueTail = &Datagram->SendQueue;
+        }
+        if (Datagram->SendQueueTail == &SendRequest->Next) {
+            Datagram->SendQueueTail = &Datagram->SendQueue;
+        }
+        Datagram->SendQueue = SendRequest->Next;
+
+        if (HadRoomForDatagram) {
+            Builder->Metadata->Flags.IsRetransmittable = TRUE; // TODO - It's ack-eliciting but not really retransmittable
+            Builder->Metadata->Frames[Builder->Metadata->FrameCount].Type = QUIC_FRAME_DATAGRAM;
+            Builder->Metadata->Frames[Builder->Metadata->FrameCount].DATAGRAM.ClientContext = SendRequest->ClientContext;
+            QuicDatagramCompleteSend(
+                Connection,
+                SendRequest,
+                &Builder->Metadata->Frames[Builder->Metadata->FrameCount].DATAGRAM.ClientContext);
+            if (++Builder->Metadata->FrameCount == QUIC_MAX_FRAMES_PER_PACKET) {
+                return TRUE;
+            }
+        } else {
+            QuicDatagramCancelSend(Connection, SendRequest);
+        }
+    }
+
+    Connection->Send.SendFlags &= ~QUIC_CONN_SEND_FLAG_DATAGRAM;
+
+    return TRUE;
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+BOOLEAN
+QuicDatagramProcessFrame(
+    _In_ QUIC_DATAGRAM* Datagram,
+    _In_ const QUIC_RECV_PACKET* const Packet,
+    _In_ QUIC_FRAME_TYPE FrameType,
+    _In_ uint16_t BufferLength,
+    _In_reads_bytes_(BufferLength)
+        const uint8_t * const Buffer,
+    _Inout_ uint16_t* Offset
+    )
+{
+    QUIC_DATAGRAM_EX Frame;
+    if (!QuicDatagramFrameDecode(FrameType, BufferLength, Buffer, Offset, &Frame)) {
+        return FALSE;
+    }
+
+    const QUIC_BUFFER QuicBuffer = { (uint16_t)Frame.Length, (uint8_t*)Frame.Data };
+
+    QUIC_CONNECTION_EVENT Event;
+    Event.Type = QUIC_CONNECTION_EVENT_DATAGRAM_RECEIVED;
+    Event.DATAGRAM_RECEIVED.Buffer = &QuicBuffer;
+    if (Packet->EncryptedWith0Rtt) {
+        Event.DATAGRAM_RECEIVED.Flags = QUIC_RECEIVE_FLAG_0_RTT;

```

Comment:
We prohibit sending datagrams with 0-rtt, but we allow receiving them as 0-rtt?

Reply:
Why do you say sending with 0-RTT is prohibited?

Reply:
[datagram.c line 347](https://github.com/microsoft/msquic/pull/329/files#diff-a43ca51fc0048a0305b2b1d4e053b62dR347)